python/extensions/pip.bzl

```diff
@@ -14,63 +14,140 @@
+load("@rules_python//python:pip.bzl", "whl_library_alias")
+load(
+    "@rules_python//python/pip_install:pip_repository.bzl",
+    "locked_requirements_label",
+    "multiple_pip_repository_bzlmod",
+    "pip_repository_attrs",
+    "pip_repository_bzlmod",
+    "use_isolated",
+    "whl_library",
+)
 load("@rules_python//python/pip_install:requirements_parser.bzl", parse_requirements = "parse")
 
+# Printing a warning msg not debugging, so we have to disable
+# the buildifier check.
+# buildifier: disable=print
+def _print_warn(msg):
+    print("WARNING:", msg)
+
+def _create_pip(module_ctx, pip_attr, hub_name, whl_map):
+    pip_name = hub_name + "_{}".format(pip_attr.python_version.replace(".", ""))
+    requrements_lock = locked_requirements_label(module_ctx, pip_attr)
+
+    # Parse the requirements file directly in starlark to get the information
+    # needed for the whl_libary declarations below. This is needed to contain
+    # the pip_repository logic to a single module extension.
+    requirements_lock_content = module_ctx.read(requrements_lock)
+    parse_result = parse_requirements(requirements_lock_content)
+    requirements = parse_result.requirements
+    extra_pip_args = pip_attr.extra_pip_args + parse_result.options
+
+    # Create the repository where users load the `requirement` macro. Under bzlmod
+    # this does not create the install_deps() macro.
+    pip_repository_bzlmod(
+        name = pip_name,
+        repo_name = pip_name,
+        requirements_lock = pip_attr.requirements_lock,
+    )
+    if hub_name not in whl_map:
+        whl_map[hub_name] = {}
+
+    for whl_name, requirement_line in requirements:
+        whl_name = _sanitize_name(whl_name)
+        whl_library(
+            name = "%s_%s" % (pip_name, whl_name),
+            requirement = requirement_line,
+            repo = pip_name,
+            repo_prefix = pip_name + "_",
+            annotation = pip_attr.annotations.get(whl_name),
+            python_interpreter = pip_attr.python_interpreter,
+            python_interpreter_target = pip_attr.python_interpreter_target,
+            quiet = pip_attr.quiet,
+            timeout = pip_attr.timeout,
+            isolated = use_isolated(module_ctx, pip_attr),
+            extra_pip_args = extra_pip_args,
+            download_only = pip_attr.download_only,
+            pip_data_exclude = pip_attr.pip_data_exclude,
+            enable_implicit_namespace_pkgs = pip_attr.enable_implicit_namespace_pkgs,
+            environment = pip_attr.environment,
+        )
+
+        if whl_name not in whl_map[hub_name]:
+            whl_map[hub_name][whl_name] = {}
+
+        whl_map[hub_name][whl_name][pip_attr.python_version] = pip_name + "_"
+
+    return whl_map
+
 def _pip_impl(module_ctx):
+    root_hubs = {}
+    submodules_hubs = {}
+    hub_whl_map = {}
+
+    # TODO we need to check for duplicate python versions in the same hub
+    # TODO we need to check for duplicate python hub names in submodules
```

Comment:
To identify name conflicts, keep a mapping of global hub names to their defining module.

Reply:
I could use a walkthrough of how your refactored toolchains. The pattern you used is sweet.

we need to check for unique names and also need to check that we don’t have duplicate pip python versions in the same hub.